File: data structure and algorithms/Depth-First Search/99.recover-binary-search-tree.java
/**
 * Algorithm:
 * Inorder traversal of a BST should be increasing. Two swapped nodes create
 * one or two inversions. The first bad node is prev in the first inversion; the
 * second bad node is current in the last inversion. Swap their values.
 *
 * Complexity:
 * Time O(n), recursion space O(h).
 */
class Solution {
    private TreeNode first;
    private TreeNode second;
    private TreeNode prev;

    public void recoverTree(TreeNode root) {
        first = null;
        second = null;
        prev = null;
        inorder(root);
        int tmp = first.val;
        first.val = second.val;
        second.val = tmp;
    }

    private void inorder(TreeNode node) {
        if (node == null) {
            return;
        }
        inorder(node.left);
        if (prev != null && prev.val > node.val) {
            if (first == null) {
                first = prev;
            }
            second = node;
        }
        prev = node;
        inorder(node.right);
    }
}

File: data structure and algorithms/Binary Tree/99.recover-binary-search-tree.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

// Definition for a binary tree node.
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class Solution {
public:
    void recoverTree(TreeNode* root) {
        /*
        Approach:
        Inorder traversal of a BST should be strictly increasing. Swapping two
        nodes creates one or two inversions. The previous node in the first
        inversion is first; the current node in the last inversion is second.
        Swap their values to restore the tree structure.

        Complexity: O(n) time and O(h) recursion space.
        */
        TreeNode* first = nullptr;
        TreeNode* second = nullptr;
        TreeNode* prev = nullptr;
        function<void(TreeNode*)> inorder = [&](TreeNode* node) {
            if (!node) return;
            inorder(node->left);
            if (prev && prev->val > node->val) {
                if (!first) first = prev;
                second = node;
            }
            prev = node;
            inorder(node->right);
        };
        inorder(root);
        if (first && second) swap(first->val, second->val);
    }
};
